### ISSUE TITLE
build issues with 2.2.7, 2.2.8 and TBB

### ISSUE THREAD

**Issue by greyltc** [BenLangmead/bowtie2#34]
Hi,
I seem to be having some trouble building with WITH_TBB=1
This is for 2.2.7, I could also not build 2.2.6 with TBB though.

[Here is what I get](https://gist.githubusercontent.com/greyltc/64e88895fd5728a4eaeb/raw/550ffbbb8c81195bff5e4638da61d828615ce15a/TBB_build.log) when I do  `WITH_TBB=1 make`
TLDR: the build is broken with numerous `error: ‘tthread’ was not declared in this scope`

and

[Here is what I get](https://gist.githubusercontent.com/greyltc/87c4c78b4df8b0ad97af/raw/bebbd8d1cf7519e71bf17c07a4ea86a327ac22a5/build.log) when I do  `WITH_TBB=0 make`
TLDR: the build goes fine

Can you give me any tips for solving this?

**Comment by aviantask** [comment]
Hi,

I have a similar issue when installing bowtie2 2.2.6 version WITH_TBB=1 via linuxbrew  , the build without TBB option goes fine.


[Here is the make log](https://github.com/BenLangmead/bowtie2/files/137022/01.make.txt)

**Comment by jdidion** [comment]
I get similar issues. I can build 2.2.6 fine. Attached is my error log.
[01.make.txt](https://github.com/BenLangmead/bowtie2/files/138418/01.make.txt)

**Comment by val-antonescu** [comment]
Yes, the TBB was not properly enabled for bowtie2-build binaries. We fixed this and will release the code by the end of this week. Thanks to all for pointing this out to us.

Val

**Comment by greyltc** [comment]
Hi @BenLangmead, @val-antonescu,
TBB build issues persist in the v2.2.8 release: [here's my build log](http://pastebin.com/raw/89kQmHLp)

My TBB is up to date.
I'm using the latest official release, 4.4 update 3 [from here] (https://www.threadingbuildingblocks.org/sites/default/files/software_releases/).

**Comment by val-antonescu** [comment]
It seems we all have version 4.1 here and unfortunately that was the version we tested with. We will try to install a 4.4 and see if this is a version related issue.

Val

**Comment by sjackman** [comment]
I have this issue as well. See https://github.com/Homebrew/homebrew-science/issues/3440 and https://github.com/Homebrew/homebrew-science/issues/3441

**Comment by sjackman** [comment]
Here's a build failure on Travis CI: https://travis-ci.org/Homebrew/homebrew-science/jobs/115917040#L786

**Comment by BenLangmead** [comment]
Thanks, all!  these are very helpful comments.  We're setting up TBB 4.4 so we can recreate and fix.

**Comment by sjackman** [comment]
Have you tried out Linuxbrew? It's very helpful for installing new versions of software. Disclaimer: I maintain Linuxbrew. See http://linuxbrew.sh
To reproduce this issue:
```sh
brew install homebrew/science/bowtie2
```

**Comment by BenLangmead** [comment]
We think the commit we just made to the master branch fixes this issue.  If you have a chance and can confirm that the current master branch compiles, that would be helpful.

**Comment by BenLangmead** [comment]
We had to make a functor member function `const`, per this item in the TBB 4.4 update 1 change log:
```
- Usage of mutable functors with task_group::run_and_wait() and
    task_arena::enqueue() is disabled. An attempt to pass a functor
    which operator()() is not const will produce compilation errors.
```

**Comment by sjackman** [comment]
https://gist.github.com/sjackman/2ea43b9f87f58ac3bff5#file-01-make-L55
```
/gsc/btl/linuxbrew/include/tbb/task.h: In instantiation of 'tbb::task* tbb::internal::function_task<F>::execute() [with F = BinarySorting_worker<S2bDnaString>]':
bt2_build.cpp:669:1:   required from here
/gsc/btl/linuxbrew/include/tbb/task.h:913:13: error: no match for call to '(const BinarySorting_worker<S2bDnaString>) ()'
             my_func();
```

You can reproduce this error in a Docker image with:
```sh
docker run linuxbrew/linuxbrew brew install --HEAD homebrew/science/bowtie2
```

**Comment by BenLangmead** [comment]
Silly me -- I had made the change on the alignment side but not on the index-building side.  Should be fixed now.  I can build all targets with WITH_TBB=1 in the master now.

**Comment by sjackman** [comment]
Yep, it works now! Ping me when there's a new release, and I'll update the Homebrew-science formula.

**Comment by greyltc** [comment]
Please notify me when there's a release also. I maintain the Arch Linux package.

**Comment by sjackman** [comment]
@BenLangmead Please tag a stable release, when you get a chance.

**Comment by val-antonescu** [comment]
I tested 2.2.9 and it seems to be working at the moment. So I will close this case for now.

thanks all for help,
Val

**Comment by sjackman** [comment]
I've submitted a PR to update Homebrew-science: https://github.com/Homebrew/homebrew-science/pull/3441

**Comment by BenLangmead** [comment]
Thanks Shuan -- let me know of any more issues.  I can successfully compile on various versions of TBB now, and have tested parallel index-building on at least a few non-trivial examples without issues.

**Comment by sjackman** [comment]
bowtie2 is merged into Homebrew-science. It compiles with recent TBB on the last three versions of Mac OS and Ubuntu 14.
